iam: Make Oauth2 service compatible with DAPS and create integration test for token retrieval/validation from a DAPS instance (closes #187)

bscholtes1A opened the pull request:
- Make JWT construction more flexible in Oauth2 service by adding a decorator registry. 
- Fix retrieval of the public key needed for verifying the token (bug in current implem)
- Add integration test issuing a request to the DAPS test instance hosted by Fraunhofer, collecting the token and then verifying that the token is valid.

ndr-brt commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/Oauth2Extension.java:
```diff
@@ -111,21 +134,40 @@ public void initialize(ServiceExtensionContext context) {
+            keyManagerFactory.init(keyStore, "test123".toCharArray());
+            sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), new SecureRandom());
+
+            return new OkHttpClient.Builder()
```
I think we should avoid to instantiate a new `OkHttpClient`, is better to create a new one starting from the one already registered using the `newBuilder()` method, so you don't have to redefine basic configurations (like `connectTimeout` and `readTimeout`)

bscholtes1A replied:
corrected

ndr-brt commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/Oauth2Extension.java:
```diff
@@ -111,21 +134,40 @@ public void initialize(ServiceExtensionContext context) {
+            KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+
+            trustManagerFactory.init(keyStore);
+            keyManagerFactory.init(keyStore, "test123".toCharArray());
```
The secret should be retrieved from the configuration or the environment variables

bscholtes1A replied:
done

ndr-brt commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/Oauth2Extension.java:
```diff
@@ -111,21 +134,40 @@ public void initialize(ServiceExtensionContext context) {
+    private KeyStore readKeyStore() {
+        try (InputStream is = getClass().getClassLoader().getResourceAsStream("keystore.jks")) {
+            KeyStore keystore = KeyStore.getInstance(KeyStore.getDefaultType());
+            String password = "test123";
```
The secret should be retrieved from the configuration or the environment variables

bscholtes1A replied:
done

ndr-brt commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/impl/IdentityProviderKeyResolver.java:
```diff
@@ -105,10 +106,10 @@ public void refreshKeys() {
         for (JwkKey key : jwkKeys) {
-            if (!"RSA".equals(key.getKty()) || !"sig".equals(key.getUse())) {
+            if (!"RSA".equals(key.getKty())) {
                 continue;
```
I would avoid the usage of the `continue` statement, better to refactor this for loop using the `Stream` approach, that `"RSA".equals(key.getKty())` condition would be simply defined as a filter.

bscholtes1A replied:
done

bscholtes1A commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/impl/Oauth2ServiceImpl.java:
```diff
@@ -173,43 +177,24 @@ public VerificationResult verifyJwtToken(String token, String audience) {
 
     @Nullable
-    private JWSVerifier createVerifier(JWSHeader header, String publicCertificateAlias) {
-        var publicKey = Objects.requireNonNull(configuration.getCertificateResolver().resolveCertificate(publicCertificateAlias)).getPublicKey();
```
Hi @paullatzelsperger ,
before I forget, this is the last point that seems weird to me in current implem:
Here for me we should be using the public key of the Authorization server (DAPS instance in case of DAPS). This one should be obtained through the PublicKeyResolver, which is not used at all in the current implem (keys are properly refreshed from the provided URL but they are never used).

Lets discuss that on Monday.

MoritzKeppler commented on extensions/iam/oauth2/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/spi/JwtDecorator.java:
```diff
@@ -0,0 +1,8 @@
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jwt.JWTClaimsSet;
+
+public interface JwtDecorator {
```
actually it's not the Decorator pattern, so the naming may be misleading - maybe "handler" is a better naming? Nevertheless I must admit I understood it intuitively.

bscholtes1A replied:
Would the term *Writer be more appropriate?

MoritzKeppler replied:
maybe *Configurator? but I'm fine with keeping Decorator for now. We can rename it as soon as we find a better name

MoritzKeppler reviewed (commented):
thx for the PR! added some findings as comments

bscholtes1A commented on extensions/iam/daps/src/main/resources/META-INF/services/org.eclipse.dataspaceconnector.spi.system.ServiceExtension:
```diff
@@ -0,0 +1,15 @@
+#  SPDX-License-Identifier: Apache-2.0
+#
+#  Contributors:
+#       Microsoft Corporation - initial API and implementation
```
Hi @MoritzKeppler 
we are in the process to get Amadeus being part of the Eclipse Fundation (this should be effective soon). What should I put there meantime?

MoritzKeppler replied:
guess you have already signed the ECA, right? This is the important step to accept contributions from you.
To my understanding you as an originator of the code can decide who owns the copyright. Is it a personal contribution or a company you are working for. Don't think that it's necessary that the company is already member of Eclipse (we also have some code copied over from other open source projects - that normally works as long as the license fits and we keep the copyright header). So I think you can add Amadeus.

bscholtes1A replied:
Yeap I signed the ECA. I put Amadeus in all headers as suggested.

bscholtes1A commented:
> thx for the PR! added some findings as comments

Hi @MoritzKeppler 
I addressed your comments. Is there any improvement to be added before we can merge this PR?

Benjamin

paullatzelsperger commented on extensions/iam/daps/src/main/java/org/eclipse/dataspaceconnector/iam/daps/DapsExtension.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Override
+    public Set<String> requires() {
+        return Set.of("oauth2");
```
we probably also require `JwtDecoratorRegistry.FEATURE`, even though the `JwtDecoratorRegistry` is covered by `"oauth2"`. just for the sake of clarity.

bscholtes1A replied:
Fixed
